hamilflow/models/harmonic_oscillator.py

```diff
@@ -63,8 +64,112 @@ class HarmonicOscillatorIC(BaseModel):
     v0: float = 0.0
 
 
-class HarmonicOscillator:
-    r"""Generate time series data for a [harmonic oscillator](https://en.wikipedia.org/wiki/Harmonic_oscillator).
+class HarmonicOscillatorBase(ABC):
+    r"""Base class to generate time series data
+    for a [harmonic oscillator](https://en.wikipedia.org/wiki/Harmonic_oscillator).
+
+    :param system: all the params that defines the harmonic oscillator.
+    :param initial_condition: the initial condition of the harmonic oscillator.
+    """
+
+    def __init__(
+        self,
+        system: Dict[str, float],
+        initial_condition: Optional[Dict[str, float]] = {},
+    ):
+        self.system = HarmonicOscillatorSystem.model_validate(system)
+        self.initial_condition = HarmonicOscillatorIC.model_validate(initial_condition)
+
+    @cached_property
+    def definition(self) -> Dict[str, float]:
+        """model params and initial conditions defined as a dictionary."""
+        return {
+            "system": self.system.model_dump(),
+            "initial_condition": self.initial_condition.model_dump(),
+        }
+
+    @abstractmethod
+    def _x(self, t: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
```

Comment:
Union[float, np.ndarray] -> https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike

Reply:
Update in 01dc720

Reply:
https://github.com/kausalflow/hamilflow/pull/48/commits/01dc720e554a580dcd534424f2987a34b224d003